cmd/promoted-image-governor/main.go

```diff
@@ -0,0 +1,255 @@
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/test-infra/prow/flagutil"
+	"k8s.io/test-infra/prow/logrusutil"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/yaml"
+
+	imagev1 "github.com/openshift/api/image/v1"
+
+	"github.com/openshift/ci-tools/pkg/api"
+	"github.com/openshift/ci-tools/pkg/config"
+	"github.com/openshift/ci-tools/pkg/steps/release"
+)
+
+type options struct {
+	ciOperatorconfigPath       string
+	kubeconfig                 string
+	dryRun                     bool
+	ignoredImageStreamTagsRaw  flagutil.Strings
+	ignoredImageStreamTags     sets.String
+	openshiftMappingDir        string
+	openshiftMappingConfigPath string
+	openshiftMappingConfig     *OpenshiftMappingConfig
+}
+
+// OpenshiftMappingConfig for openshift image mapping files
+type OpenshiftMappingConfig struct {
+	SourceRegistry  string              `json:"source_registry"`
+	TargetRegistry  string              `json:"target_registry"`
+	SourceNamespace string              `json:"source_namespace"`
+	TargetNamespace string              `json:"target_namespace"`
+	Images          map[string][]string `json:"images,omitempty"`
+}
+
+func loadOpenshiftMappingConfig(path string) (*OpenshiftMappingConfig, error) {
+	data, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read file: %s", path)
+	}
+	openshiftMappingConfig := &OpenshiftMappingConfig{}
+	if err = yaml.Unmarshal(data, openshiftMappingConfig); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal the config %q: %w", string(data), err)
+	}
+	return openshiftMappingConfig, nil
+}
+
+func newOpts() *options {
+	opts := &options{}
+	fs := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
+	// Controller-Runtimes root package imports the package that sets this flag
+	kubeconfigFlagDescription := "Path to the kubeconfig file to use for CLI requests."
+	if f := fs.Lookup("kubeconfig"); f != nil {
+		f.Usage = kubeconfigFlagDescription
+		// https://i.kym-cdn.com/entries/icons/original/000/018/012/this_is_fine.jpeg
+		defer func() { opts.kubeconfig = f.Value.String() }()
+	} else {
+		fs.StringVar(&opts.kubeconfig, "kubeconfig", "", kubeconfigFlagDescription)
+	}
+	fs.StringVar(&opts.ciOperatorconfigPath, "ci-operator-config-path", "", "Path to the ci operator config")
+	fs.BoolVar(&opts.dryRun, "dry-run", true, "Whether to run the controller-manager with dry-run")
+	fs.Var(&opts.ignoredImageStreamTagsRaw, "ignored-image-stream-tags", "An imagestreamtag to ignore while checking. It must be in namespace/name:tag format (e.G `ci/clonerefs:latest`). Can be passed multiple times.")
+	fs.StringVar(&opts.openshiftMappingDir, "openshift-mapping-dir", "", "Path to the openshift mapping directory")
+	fs.StringVar(&opts.openshiftMappingConfigPath, "openshift-mapping-config", "", "Path to the openshift mapping config file")
+	if err := fs.Parse(os.Args[1:]); err != nil {
```

Comment:
parsing, loading, etc, does not belong in the constructor for the options

Reply:
Renamed the function name. Move the validation logic to its own function.